examples/pytorch/mock_sparse/gat/train.py

```diff
@@ -0,0 +1,143 @@
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+import dgl
+from dgl.data import CoraGraphDataset, CiteseerGraphDataset, PubmedGraphDataset
+from dgl import AddSelfLoop
+import argparse
+
+from dgl.mock_sparse import create_from_coo, softmax, bspmm
+from torch.nn import init
+
+
+class GATConv(nn.Module):
+    def __init__(self, in_size, out_size, n_heads):
+        super(GATConv, self).__init__()
+        self.in_size = in_size
+        self.out_size = out_size
+        self.n_heads = n_heads
+        self.W = nn.Parameter(torch.Tensor(in_size, out_size * n_heads))
+        self.a_l = nn.Parameter(torch.Tensor(1, n_heads, out_size))
+        self.a_r = nn.Parameter(torch.Tensor(1, n_heads, out_size))
+        self.leaky_relu = nn.LeakyReLU(0.2)
+        init.xavier_uniform_(self.W)
+        init.xavier_uniform_(self.a_l)
+        init.xavier_uniform_(self.a_r)
+
+    def forward(self, A, h):
+        Wh = (h @ self.W).view(
+            -1, self.n_heads, self.out_size
+        )  # |V| x N_h x D_o
+        Wh1 = (Wh * self.a_l).sum(2)  # |V| x N_h
+        Wh2 = (Wh * self.a_r).sum(2)  # |V| x N_h
+        Wh1 = Wh1[A.row, :]
+        Wh2 = Wh2[A.col, :]
```

Comment:
Could you add shape comment to each line to make it easier to understand?

Reply:
Done